### PR TITLE
Initial spin-button a11y commit

### PR DESCRIPTION
**Pull request by jxlz**
Keyboard state diagram, and open questions on accessibility interaction and specification highlighted.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `189.057 kB`<br />`52.092 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            |  `31.883 kB`<br />`10.516 kB` |
| <samp>react-spinbutton</samp> <br /> <abbr title='bundle-size/SpinButton.fixture.js'>SpinButton</abbr>                                                                        |  `44.227 kB`<br />`12.492 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/8d120219d3605abf8feff95eaff52ab01851618e'>8d120219d3605abf8feff95eaff52ab01851618e</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/Avatar-mount_base.html">1564</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/Avatar-mount.html">1556</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/Button-mount_base.html">1068</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/Button-mount.html">1073</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/FluentProvider-mount_base.html">2129</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/FluentProvider-mount.html">2101</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/FluentProviderWithTheme-mount_base.html">721</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/FluentProviderWithTheme-mount.html">733</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">696</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">702</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">767</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">760</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/MakeStyles-mount_base.html">2290</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/MakeStyles-mount.html">2244</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/SpinButton-mount_base.html">3063</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24075/perf-test-react-components/SpinButton-mount.html">3133</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+| Committed  | Any Edit Key (that results in a change) | Editing |
+| Editing  | Any Edit Key | Editing |
+| Content after spin button (Value committed) | Shift + Tab | Committed |
+
+##### Edit keys
+| Edit Key | Result |
+| ---------- | ------------ |
+| Home | First item in defined range |
+| End | Last item in defined range |
+| Up arrow | One increment higher |
```
Comment (spmonahan) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:32]: The `step` prop controls the increment/decrement amount. It defaults to `1` but it could be `10`. What do you think about changing the results cell to

```
`step` increments higher
```

or

```
Increments by small about. Defaults to `1` but can be changed with `step` prop.
```

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+| Down arrow | One increment lower |
+| Page up | Several increments higher |
```
Comment (spmonahan) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:34]: Like the up/down arrows this value is controlled by `stepPage`. It also defaults to `1` but can be set to any numeric value.

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+| Typing valid value | Valid value |
+| Begin typing valid value | Autofill option provided |
+
+### High contrast mode
+
+Be careful about visibility of the spin button arrows in high contrast mode.
+
+### Motion and animation
+
+Animations occur when the value changes to give the appearance of spinning. This needs to respond to user settings to turn off animations.
```
Comment (smhigley) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:45]: This isn't currently part of our v9 (or v8) SpinButton. Unless we have a plan to add this, I think we can leave this line off.

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+Placeholder: Mouse press speed of value changes visual indicators
+
+### Semantics
+
+## Known issues
+
+Placeholder: If there are any current known bugs with browsers or AT that impact this component, list them here.
+
+Mouse press quick changes to value are not announced to screen readers.
```
Comment (smhigley) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:55]: This can go in a different PR, but I'm dropping this here for reference:

- NVDA doen't read the value of spinbuttons in Chromium: https://github.com/nvaccess/nvda/issues/13195
- Narrator defaults to reading min/max as 0 when they're (intentionally) undefined: https://microsoft.visualstudio.com/Edge/_workitems/edit/39070743

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+
+### Advanced usage
+
+Placeholder: Cover potential use cases not included in our storybook examples here (or even included storybook examples, if they're complex require a notable amout of work or nuanced understanding from authors)
+
+#### Child content restrictions
+
+The following child content is allowed:
+1. Text
```
Comment (smhigley) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:82]: I believe we actually do not allow any children, since the primary slot is an `<input>` element.

We could mention that the up/down buttons can have children, though.

File: packages/react-components/react-spinbutton/src/stories/SpinButton/index.stories.tsx [hunk 6]
```diff
@@ -20,7 +21,7 @@ const meta: Meta = {
   parameters: {
     docs: {
       description: {
-        component: [descriptionMd, bestPracticesMd].join('\n'),
+        component: [accessibilityMd, descriptionMd, bestPracticesMd].join('\n'),
```
Comment (smhigley) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/index.stories.tsx:24]: As much as I think a11y is important, I think the component description (and possibly the more concise best practices) should go before the a11y docs.
```suggestion
        component: [descriptionMd, bestPracticesMd, accessibilityMd].join('\n'),
```

**Review by smhigley — COMMENTED** [review]
I left a bunch of comments, though not all of them need to be addressed in this PR -- we can add future PRs with more specifics to flesh out sections.

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+| Edit Key | Result |
+| ---------- | ------------ |
+| Home | First item in defined range |
+| End | Last item in defined range |
+| Up arrow | One increment higher |
+| Down arrow | One increment lower |
+| Page up | Several increments higher |
+| Page down | Several increments lower |
+| Typing valid value | Valid value |
+| Begin typing valid value | Autofill option provided |
```
[the remark this answers is not in the document]
Comment (jxlz) [reply packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:37]: It's in the [Figma Spec](https://www.figma.com/proto/4ErMjZ0Mi72legBtMAbbEp/SpinButton?node-id=309%3A57799). We could try to discourage non-numeric numbers by removing that section.

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md (hunk 5 above)
Comment (jxlz) [reply packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:82]: Other than carats, what other children are allowed?

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+### High contrast mode
+
+Be careful about visibility of the spin button arrows in high contrast mode.
```
[the remark this answers is not in the document]
Comment (jxlz) [reply packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:41]: Todo added in the file

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (spmonahan) [reply packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md]: We intentionally omitted auto complete functionality for now because, as Sarah noted, it's not clearly a valuable feature and  it's worth considering a more general-purpose auto complete implementation that can be used across different types of inputs.

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md (hunk 5 above)
Comment (spmonahan) [reply packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md]: > Other than carats, what other children are allowed?

Technically, anything that can be a child of a `<button>`. In practice it should probably be restricted to a 16px icon from `@fluentui/react-icons` as other content will likely break the visual design.

File: change/@fluentui-react-spinbutton-803b9a0b-82dd-471f-b57d-7717eb0d925d.json [hunk 9]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "Merge branch 'master' of https://github.com/microsoft/fluentui into spin-button",
```
Comment (spmonahan) [inline change/@fluentui-react-spinbutton-803b9a0b-82dd-471f-b57d-7717eb0d925d.json:3]: This comment will be part of the public changelog so it needs to be something that make sense in that context. What do you think about?
```suggestion
  "comment": "docs: add accessibility documentation for SpinButton",
```

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+When a user holds down an edit key or mouse click, there is an animation of spinning numbers to provide a sense of how fast the value is changing. This animation is an interaction cue and therefore is acceptable even when a user has set a preference for reduced motion.
+
+### Semantics
+
+## Known issues
+
+Placeholder: If there are any current known bugs with browsers or AT that impact this component, list them here.
+
+Mouse press quick changes to value are not announced to screen readers.
+
```
Comment (spmonahan) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:55]: Some known issues.

### Narrator + Edge

1. When no `min` or `max` values are set Narrator announces "minimum 0" and "maximum 0". This is misleading because when these values are not set `SpinButton` does not enforce a min or max value. [Example](https://codesandbox.io/s/sample-spinbutton-zxy72x).

### NVDA + Edge/Chrome

1. When focused on the `SpinButton` input field pressing up/down arrows announces "blank". NVDA + Firefox announces the correct value.

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+SpinButtons are a better choice than Input and Input with type: number when clear indication is needed that there are maximum and minimum allowed values.
+
+SpinButtons are a better choice than Slider when there are many valid values and Slider would not provide enough granularity to choose a precise value.
+
+### Required props
+
+Placeholder: If there are any props or other specific work that the author must do to make the default usage of the component accessible, list those here. One example is that all form components must have a label provided by the author. If the component has any internally defined strings that should be localized, list those here.
+
+Open question: How does everything read? Name, role, value? What about when there are prefixes or postfixes?
```
Comment (spmonahan) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:72]: How the SpinButton is read depends on the screen reader used but generally it reads (see comment above about known issues):

1. Label
2. "spin button"
3. "value" where "value" is the current displayed value

> What about when there are prefixes or postfixes?

Prefixes, postfixes and other display formatted (e.g., mapping the numbers 1-12 to January-December) is handled by setting the `aria-valuetext` attribute.

How this is read also depends on the screen reader used but a screen reader should read the displayed text consistently with how it would read it in other contexts (i.e., "$1" should be read as "one dollar"). [Example](https://react.fluentui.dev/?path=/docs/components-spinbutton--default#display-value).

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+SpinButtons allow someone to incrementally adjust a value in small steps.
+
+SpinButtons are a better choice than Input and Input with type: number when clear indication is needed that there are maximum and minimum allowed values.
+
+SpinButtons are a better choice than Slider when there are many valid values and Slider would not provide enough granularity to choose a precise value.
+
+### Required props
+
+Placeholder: If there are any props or other specific work that the author must do to make the default usage of the component accessible, list those here. One example is that all form components must have a label provided by the author. If the component has any internally defined strings that should be localized, list those here.
```
Comment (spmonahan) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:70]: Authors must provide a label for `SpinButton`. The recommended pattern for Fluent inputs is to use the [`Label`](https://react.fluentui.dev/?path=/docs/components-label--default) component like this:

```jsx
<Label htmlFor="example-spinbutton">A SpinButton</Label>
<SpinButton id="example-spinbutton" defaultValue={10} />
```

[Here's a live example](https://codesandbox.io/s/sample-spinbutton-zxy72x?file=/src/index.tsx:532-639).

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+### Advanced usage
+
+Placeholder: Cover potential use cases not included in our storybook examples here (or even included storybook examples, if they're complex require a notable amout of work or nuanced understanding from authors)
+
+#### Child content restrictions
+
+The component has the following structure:
+* Input field
+* Up button
+  * Up carat symbol
```
Comment (spmonahan) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:87]: What do you think about?
```suggestion
  * Increment icon
```

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+* Down button
+  * Down carat symbol
```
Comment (spmonahan) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:89]: and
```suggestion
  * Decrement icon
```

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+
+#### Component-specific usage warnings
```
Comment (spmonahan) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:92]: These seem like they belong in the "Known Issues" section.

**Review by spmonahan — COMMENTED** [review]
Looking good!

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 16]
```diff
@@ -49,10 +49,18 @@ When a user holds down an edit key or mouse click, there is an animation of spin
...
-Placeholder: If there are any current known bugs with browsers or AT that impact this component, list them here.
-
 Mouse press quick changes to value are not announced to screen readers.
 
+[NVDA does not read the value of spinbuttons in Chromium](https://github.com/nvaccess/nvda/issues/13195)
+
+[Narrator defaults to reading min/max as 0 when they're (intentionally) undefined](https://microsoft.visualstudio.com/Edge/_workitems/edit/39070743)
+
+### Narrator + Edge
+When no min or max values are set Narrator announces "minimum 0" and "maximum 0". This is misleading because when these values are not set SpinButton does not enforce a min or max value. Example.
```
Comment (spmonahan) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:59]:
```suggestion
When no min or max values are set Narrator announces "minimum 0" and "maximum 0". This is misleading because when these values are not set SpinButton does not enforce a min or max value.
```

File: packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md [hunk 17]
```diff
@@ -67,9 +75,21 @@ SpinButtons are a better choice than Slider when there are many valid values and
...
+```
+
+How the SpinButton is read depends on the screen reader used but generally it reads:
+
+1. Label
+2. "spin button"
+3. "value" where "value" is the current displayed value
 
-Open question: How does everything read? Name, role, value? What about when there are prefixes or postfixes?
+Prefixes, postfixes and other display formatted (e.g., mapping the numbers 1-12 to January-December) is handled by setting the `aria-valuetext` attribute.
```
Comment (spmonahan) [inline packages/react-components/react-spinbutton/src/stories/SpinButton/SpinButtonAccessibility.md:90]:
```suggestion
Prefixes, postfixes and other display formatting (e.g., mapping the numbers 1-12 to January-December) is handled by setting the `displayValue` prop which, internally to `SpinButton` sets the `aria-valuetext` attribute.
```

**Review by spmonahan — APPROVED** [review]
Looks good! Just a few wording suggestions.

**Review by spmonahan — APPROVED** [review]
